stdlib/Scratch/docs/index.md

```diff
@@ -0,0 +1,145 @@
+
+## API overview
+
+Scratch space usage is performed primarily through one function: `get_scratch!()`.
+It provides a single interface for creating and getting previously-created spaces, either tied to a package by its UUID, or as a global scratch space that can be accessed by any package.
+Here is an example where a package creates a scratch space that is namespaced to its own UUID:
+
+```julia
+module ScratchExample
+using Scratch
+
+# This will be filled in inside `__init__()`
+download_cache = ""
+
+# Downloads a resource, stores it within a scratchspace
+function download_dataset(url)
+    fname = joinpath(download_cache, basename(url))
+    if !isfile(fname)
+        download(url, fname)
+    end
+    return fname
+end
+
+function __init__()
+    global download_cache = @get_scratch!("downloaded_files")
+end
+
+end # module ScratchExample
+```
+
+Note that we initialize the `download_cache` within `__init__()` so that our packages are as relocatable as possible; we typically do not want to bake absolute paths into our precompiled files.
+This makes use of the `@get_scratch!()` macro, which is identical to the `get_scratch!()` method, except it automatically determines the UUID of the calling module, if possible. The user can manually pass in a `Module` as well for a slightly more verbose incantation:
+```julia
+function __init__()
+    global download_cache = get_scratch!("downloaded_files", @__MODULE__)
+end
+```
+
+If a user wishes to manually delete a scratch space, the method `delete_scratch!(key; pkg_uuid)` is the natural analog to `get_scratch!()`, however in general users will not need to do so, the scratch spaces will be garbage collected by `Pkg.gc()` automatically.
+
+For a full listing of docstrings and methods, see the [Scratch Space Reference](@ref) section.
+
+
+## Tips and Tricks
+
+> Can I trigger data regeneration if the scratch space is found to be empty/files are missing?
+
+Yes, this is quite simple; just check the contents of the directory when you first call `get_scratch!()`, and if it's empty, run your generation function:
+
+```julia
+using Scratch
+
+function get_dataset_dir()
+    dataset_dir = @get_scratch!("dataset")
+    if isempty(readdir(dataset_dir))
+        perform_expensive_dataset_generation(dataset_dir)
+    end
+    return dataset_dir
+end
+```
+
+This ensures your package is resilient against situations such as scratch spaces being deleted by a user that has called `clear_scratchspaces!()` to free up disk space.
+
+> Can I create a scratch space that is not shared across versions of my package?
+
+Yes!  Make use of the `key` parameter and Pkg's ability to look up the current version of your package at compile-time:
+
+```julia
+module VersionSpecificExample
+using Pkg.TOML, Scratch
+
+# Get the current version at compile-time, that's fine it's not going to change. ;)
+function get_version()
```

Comment:
This is a very useful function in its own. By putting in the official docs, does this become the official way to query for the own package version at compile-time?

Reply:
I think you'd have to ask Fredrik or Kristoffer.  I had some vague ideas on how to do this properly, but it seems there are edge cases I didn't take into account.  The text above this assumed that my previous Pkg PR got merged, which it didn't, so I'll fix that now.